### PR TITLE
use color.shadeXxx instead of color[Xxx]

### PR DESCRIPTION
**Pull request by a14n**

### CODE REVIEW

File: dev/benchmarks/complex_layout/lib/main.dart [hunk 1]
```diff
@@ -633,7 +633,7 @@ class FancyDrawerHeader extends StatelessWidget {
   Widget build(BuildContext context) {
     return new Container(
       decoration: new BoxDecoration(
-        backgroundColor: Colors.purple[500]
+        backgroundColor: Colors.purple.shade500
```
Comment (abarth) [inline dev/benchmarks/complex_layout/lib/main.dart]: For shade500, you can just skip it and say `Colors.purple`

**Comment by abarth** [comment]
Looks good except for the 500 optimization

**Comment by Hixie** [comment]
What is the purpose of this change?

**Comment by sethladd** [comment]
As a user, I would find it useful, when reading through the code, to see the recommended way to reference a color. (assuming this new API is recommended)

**Comment by abarth** [comment]
IMHO, our examples should all three patterns: (1) Using Colors.foo directly, where appropriate, (2) using shadeX, (3) using operator[].

**Comment by Hixie** [comment]
@sethladd There's no particular recommended way. Both styles are fine.

**Comment by Hixie** [comment]
I would be fine with the examples using a mixture.

**Comment by sethladd** [comment]
> Both styles are fine.

Curious why we have two styles? I'm assuming the two styles are equivalent, but maybe that's not correct.

If possible, eliminating the question of "which style should I use? they are functionally identical" is helpful for our developers. I can imagine new developers ask "why are there two ways? what don't I know? which should I use?"

If there's a way to document "use style A when you need to do B, and use style X when you need to do style Y", that would be helpful. Looking to help reduce seemingly arbitrary choices like this, adding some way to help our developers make an informed choice.

**Comment by a14n** [comment]
I find `Colors.foo[xxx]` error prone. If you make a typo like `Colors.foo[80]` instead of  `Colors.foo[800]` you will only catch it at runtime. `Colors.foo80` is on the other side catched by the analyzer.

Moreover as there's a limited valid indexes for [MaterialColor](https://github.com/flutter/flutter/blob/fac2fac1d67708da14364dc90f40e06130da9d21/packages/flutter/lib/src/material/colors.dart#L43-L45) and [MaterialAccentColor](https://github.com/flutter/flutter/blob/fac2fac1d67708da14364dc90f40e06130da9d21/packages/flutter/lib/src/material/colors.dart#L67-L70) I've wondered if we should deprecate the use of `operator []` on them. wdyt ?

**Comment by abarth** [comment]
LGTM but travis is unhappy.

**Comment by abarth** [comment]
Looks bogus.  Re-running.

**Comment by Hixie** [comment]
The argument that it's better for static checking is a good one.

> I've wondered if we should deprecate the use of operator [] on them. wdyt ?

We don't deprecate. We either remove, or support.

There's reason to have the indexing, for example so you can loop through the colours. So I think we should keep them. It might make sense to expose the list of keys, to aid with that.

**Comment by a14n** [comment]
The test failed because of asymetry in equality : `Colors.blue == Colors.blue.shade500` is `true` but `Colors.blue.shade500 == Colors.blue` is `false`.

I can easily fix the test by adding a `.shade500` but this asymetry is not really cool.

**Comment by Hixie** [comment]
The asymmetry is a a bug in `Color.==` (it should use runtimeType). Fixing this will make it return false. A MaterialColor is not the same as the Color of the same value, since the MaterialColor has a whole map and so on.

**Pull request merged**